Libraries/LibJS/Runtime/ArrayPrototype.cpp

```diff
@@ -347,6 +348,54 @@ Value ArrayPrototype::index_of(Interpreter& interpreter)
     return Value(-1);
 }
 
+Value ArrayPrototype::reduce(Interpreter& interpreter)
+{
+    auto this_value = interpreter.this_value();
+    auto* this_object = this_value.to_object(interpreter);
+    if (!this_object)
+        return {};
+
+    auto initial_length = get_length(interpreter, *this_object);
+    if (interpreter.exception())
+        return {};
+
+    size_t start = 1;
+
+    auto accumulator = js_undefined();
+    if (interpreter.argument_count() > 1) {
+        accumulator = interpreter.argument(1);
+        start = 0;
+    } else if (initial_length >= 1) {
+        accumulator = this_object->get_by_index(0);
+    } else {
+        interpreter.throw_exception<TypeError>(String::format("Reduce of empty array with no initial value"));
+        return {};
+    }
+
+    auto* callback_function = callback_from_args(interpreter, "reduce");
+    if (!callback_function)
+        return {};
+
+    for (size_t i = start; i < initial_length; ++i) {
+        auto value = this_object->get_by_index(i);
+        if (value.is_empty()) {
+            continue;
+        }
+
+        MarkedValueList arguments(interpreter.heap());
+        arguments.append(accumulator);
+        arguments.append(value);
+        arguments.append(Value((i32)i));
+        arguments.append(this_object);
+
+        accumulator = interpreter.call(*callback_function, this_value, move(arguments));
```

Comment:
The *this* value here seems to be incorrect, the spec says:

> Set accumulator to ? Call(callbackfn, undefined, « accumulator, kValue, k, O »).

Reply:
Yes, it looks like it should be just `js_undefined()` for `reduce` although V8 just leaves the function binding as it is. I will implement what the spec suggests though.

Reply:
> although V8 just leaves the function binding as it is.

Looks like both SpiderMonkey and V8 do use `undefined` as this value:

```
> [1, 2, 3].reduce(function() { "use strict"; this.foo = "bar" })
Uncaught TypeError: Cannot set property 'foo' of undefined
    at repl:1:54
    at Array.reduce (<anonymous>)
```

Reply:
Ah! They do in strict mode! Right, so we're in strict by default?

Reply:
No strict mode in LibJS yet - that was just for demonstration purposes, as in non-strict mode this is still `undefined` here but assigning a property to it just fails silently :)

In this PR's implementation it would assign the property to the array that `reduce` is being called on:

```
> a = [1, 2, 3]
[ 1, 2, 3 ]
> a.reduce(() => { this.foo = "bar" })
> a.foo
"bar"
```